File: gifts.gl2/gl_tutor/explore/ramp.c
	/* this program creates a color ramp of 240 shades of red
		and displays them with their color indices written in white */
#include "gl.h"
#include "device.h"

/* set up some constants */

#define BASEC	8	/* the base color index */
#define ENDC	248		/* the end color index */
#define NXBOXES	24		/* number of boxes in the x direction */
#define NYBOXES	10		/* number of boxes in the y direction */
#define SIZE	(XMAXSCREEN / NXBOXES) /* size (in pixels) of each box */

char buf[100];		/* buf temporarily holds the value of color index
			variable so it can be printed by charstr */

main()
{
	initialize();
	while(TRUE) {
		drawimage();
		checkinput();
		}
}


initialize()
{
	/* declare i as an integer variable */
	int i;

	prefposition(0, 1023, 0, 767);
	winopen("color.c");

	/* load color map locations 8-247 with RGB values of
		increasing red component */
	for (i=BASEC; i<ENDC; i=i+1)
		mapcolor(i, i, 0, 0);
}


drawimage()
{
	int i,j;

	color(BLACK);
	clear();

	for (i=0; i<NYBOXES; i++)
		for (j=0; j<NXBOXES; j++){
			color(i*NXBOXES+j+BASEC);
			rectfi(j*SIZE, i*SIZE, (j+1)*SIZE, (i+1)*SIZE);
			color(WHITE);
			cmov2i(5+j*SIZE, 5+i*SIZE);
			sprintf(buf,"%d ",i*NXBOXES+j+BASEC);
			charstr(buf);
			}
}

checkinput()
{
	Device val;

	switch(qread(&val)) {
		case REDRAW:
			reshapeviewport();
			break;
		}
}

File: gifts.gl2/fgl_tutor/explore/ramp.f
C -------------------------- PROGRAM RAMP ---------------------------
C This program creates a color ramp of 240 shades of red
C and displays them with their color indices written in white.
C -------------------------------------------------------------------

$INCLUDE /usr/include/fgl.h
$INCLUDE /usr/include/fdevice.h

C -------------------------------------------------------------------
C Set up constants for the base and end color index, the number of
C boxes in the x direction, the number of boxes in the y direction,
C and the size (in pixels) of each box.
C -------------------------------------------------------------------
       INTEGER BASEC, ENDC, NXBOX, NYBOX, SIZE
       COMMON /GLBLS/BASEC, ENDC, NXBOX, NYBOX, SIZE

       BASEC = 8
       ENDC = 248
       NXBOX = 24
       NYBOX = 10
       SIZE = XMAXSC/NXBOX

       CALL INITIALIZE
 100   CALL DRAWIMAGE
       CALL CHECKINPUT
       GOTO 100

       END


       SUBROUTINE INITIALIZE


$INCLUDE /usr/include/fgl.h
$INCLUDE /usr/include/fdevice.h

       INTEGER BASEC, ENDC, NXBOX, NYBOX, SIZE
       COMMON /GLBLS/BASEC, ENDC, NXBOX, NYBOX, SIZE

C -------------------------------------------------------------------
C      Declare i as an integer variable.
C -------------------------------------------------------------------
      INTEGER I
      INTEGER WINID

       CALL PREFPO(0, 1023, 0, 767)
       WINID = WINOPE('ramp.f',6)

C -------------------------------------------------------------------
C Load color map locations 8-247 with RGB values of increasing
C red component.
C -------------------------------------------------------------------
       DO 200 I = BASEC, ENDC-1, 1
 200   CALL MAPCOL(I, I, 0, 0)

       RETURN
       END


       SUBROUTINE DRAWIMAGE

$INCLUDE /usr/include/fgl.h
$INCLUDE /usr/include/fdevice.h

       INTEGER BASEC, ENDC, NXBOX, NYBOX, SIZE
       COMMON /GLBLS/BASEC, ENDC, NXBOX, NYBOX, SIZE

       INTEGER I,J

C -------------------------------------------------------------------
C buf temporarily holds the value of the color index variable so
C charst can print it.
C -------------------------------------------------------------------
       CHARACTER*100  BUF

       CALL COLOR(BLACK)
       CALL CLEAR

       DO 400 I = 0, NYBOX-1, 1
           DO 300 J = 0, NXBOX-1, 1
		CALL COLOR(I*NXBOX + J + BASEC)
		CALL RECTFI(J*SIZE, I*SIZE, (J+1)*SIZE, (I+1)*SIZE)
		CALL COLOR(WHITE)
		CALL CMOV2I(5 + J*SIZE, 5 + I*SIZE)
		WRITE(BUF, '(I3)') I * NXBOX + J + BASEC
		CALL CHARST(BUF, 3)
 300       CONTINUE
 400   CONTINUE

       RETURN
       END


       SUBROUTINE CHECKINPUT


$INCLUDE /usr/include/fgl.h
$INCLUDE /usr/include/fdevice.h

       INTEGER BASEC, ENDC, NXBOX, NYBOX, SIZE
       COMMON /GLBLS/BASEC, ENDC, NXBOX, NYBOX, SIZE

       INTEGER*2 VAL

       IF (QREAD(VAL) .EQ. REDRAW) THEN
           CALL RESHAP
       ENDIF

       RETURN
       END
